src/java/org/apache/cassandra/utils/btree/BTree.java

```diff
@@ -1448,4 +1920,2355 @@ private static boolean isStopSentinel(long v)
+            // invariants:  iub > unode[upos], uub > inode[ipos]
+            // invariants:  ipos >=  0 && ipos  < isz (imposed by SimpleTreeIterator)
+            // invariants:  upos >= -1 && upos <= usz
+            while (true)
+            {
+                if (level == leaf())
+                {
+                    if (isLeaf(inode))
+                    {
+                        // Both are leaves, so can simply linearly merge them until we reach the update upper bound.
+                        // This is most easily achieved by looping until one of the leaves is consumed, then
+                        // continuing to consume inode until we hit uub
+
+                        // Note that unode can already be exhausted, because we use the original tree to define the
+                        // shape of our updated tree and so we retain the _bounds_ of each sub-tree, only creating
+                        // new bounds during overflow, so once we have copied a leaf we continue inserting elements
+
+                        Insert ik =  (Insert) inode[ipos];
+                        if (upos < usz)
+                        {
+                            Existing uk = (Existing) unode[upos];
+                            int c = comparator.compare(uk, ik);
+                            while (true)
+                            {
+                                if (c == 0)
+                                {
+                                    leaf().addKey(updateF.apply(uk, ik));
+                                    if (++upos < usz)
+                                        uk = (Existing) unode[upos];
+                                    if (++ipos < isz)
+                                        ik = (Insert) inode[ipos];
+                                    if (upos == usz || ipos == isz)
+                                        break;
+                                    c = comparator.compare(uk, ik);
+                                }
+                                else if (c < 0)
+                                {
+                                    int ulim = exponentialSearch(comparator, unode, upos + 1, usz, ik);
+                                    c = ulim < 0 ? 1 : 0; // positive or zero
+                                    if (ulim < 0) ulim = -(1 + ulim);
+                                    leaf().copy(unode, upos, ulim - upos);
+                                    if ((upos = ulim) == usz)
+                                        break;
+                                    uk = (Existing) unode[upos];
+                                }
+                                else
+                                {
+                                    int ilim = exponentialSearch(comparator, inode, ipos + 1, isz, uk);
+                                    c = ilim & 0x80000000; // negative or zero
```

Comment:
Let's extract this to a (static) function called e.g. `comparisonFromSearchResult` implemented as `x -> x >> Integer.BITS - 1`, so that we can use `comparisonFromSearchResult(ilim)` here and `-comparisonFromSearchResult(ilim)` above.

Reply:
wfm